powerbi-visual-builder/src_visual/visual.ts

```diff
@@ -0,0 +1,480 @@
+                  const highlight =
+                    valueColumn.highlights && // Highlights exists
+                    valueColumn.highlights[i] != null && // Not null
+                    valueColumn.highlights[i].valueOf() == value.valueOf(); // Same value
+                  if (!highlight) {
+                    allHighlighted = false;
+                  }
+                  if (
+                    column.metadata &&
+                    column.metadata.kind == "categorical"
+                  ) {
+                    obj[column.powerBIName] = obj[
+                      column.powerBIName
+                    ].toString();
+                  }
+                }
+                rowInfo.set(obj, {
+                  highlight: allHighlighted,
+                  index: i
+                });
+                return obj;
+              })
+              .filter(x => x != null)
+          }
+        ]
+      };
+      return { dataset, rowInfo };
+    }
+
+    protected getProperties(options: VisualUpdateOptions) {
+      const defaultProperties: { [name: string]: any } = {};
+      for (const p of this.template.properties as PowerBIProperty[]) {
+        defaultProperties[p.powerBIName] = p.default;
+      }
+
+      if (
+        !options ||
+        !options.dataViews ||
+        !options.dataViews[0] ||
+        !options.dataViews[0].metadata
+      ) {
+        return defaultProperties;
+      }
+      const objects = options.dataViews[0].metadata.objects;
+      if (!objects) {
+        return defaultProperties;
+      }
+
+      for (const p of this.template.properties as PowerBIProperty[]) {
+        const object = objects.chartOptions;
+        if (object[p.powerBIName] != undefined) {
+          defaultProperties[p.powerBIName] = object[p.powerBIName];
+        }
+      }
+
+      return defaultProperties;
+    }
+
+    protected unmountContainer() {
+      if (this.chartContainer != null) {
+        this.chartContainer.unmount();
+        this.chartContainer = null;
+      }
+    }
+
+    public update(options: VisualUpdateOptions) {
+      this.properties = this.getProperties(options);
+      runAfterInitialized(() => this.updateRun(options));
+    }
+
+    protected updateRun(options: VisualUpdateOptions) {
+      try {
+        this.resize(options.viewport.width, options.viewport.height);
+
+        const getDatasetResult = this.getDataset(options);
+        if (getDatasetResult == null) {
+          // If dataset is null, show a warning message
+          this.divChart.innerHTML = `
+                <h2>Dataset incomplete. Please specify all data fields.</h2>
+            `;
+          this.currentDatasetJSON = null;
+          this.unmountContainer();
+        } else {
+          const { dataset } = getDatasetResult;
+          // Check if dataset is the same
+          const datasetJSON = JSON.stringify(dataset);
+          if (datasetJSON != this.currentDatasetJSON) {

```

Comment:
This could be terribly slow when you have 5k+ items, cause `updateRun` will be called while the visual is being resized.  It might be helpful to use this property to https://github.com/Microsoft/PowerBI-visuals-tools/blob/master/templates/visuals/.api/v2.1.0/PowerBI-visuals.d.ts#L1407 to not do this check for resizing updates.

Reply:
Actually, that might be better as a separate issue, since line wasn't really introduced in this PR.

Reply:
Good catch! Let's fix this in another pull request. Since charticulator is not very fast at resizing, we probably want to add a throttle for resize as well.